pkg/util/validation/user_limits_handler.go

```diff
@@ -0,0 +1,75 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/cortexproject/cortex/blob/master/pkg/util/validation/user_limits_exposition.go
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: The Cortex Authors.
+
+package validation
+
+import (
+	"net/http"
+	"time"
+
+	"github.com/grafana/dskit/tenant"
+
+	"github.com/grafana/mimir/pkg/util"
+)
+
+type UserLimitsResponse struct {
+	CompactorBlocksRetentionPeriod float64 `json:"compactor_blocks_retention_period_second"` // suffix with second to make it explicit the value is in second
+
+	// Write path limits
+	IngestionRate             float64 `json:"ingestion_rate"`
+	IngestionBurstSize        int     `json:"ingestion_burst_size"`
+	MaxGlobalSeriesPerUser    int     `json:"max_global_series_per_user"`
+	MaxGlobalSeriesPerMetric  int     `json:"max_global_series_per_metric"`
+	MaxGlobalExemplarsPerUser int     `json:"max_global_exemplars_per_user"`
+
+	// Read path limits
+	MaxChunksPerQuery            int `json:"max_fetched_chunks_per_query"`
+	MaxFetchedSeriesPerQuery     int `json:"max_fetched_series_per_query"`
+	MaxFetchedChunkBytesPerQuery int `json:"max_fetched_chunk_bytes_per_query"`
+
+	// Ruler limits
+	RulerMaxRulesPerRuleGroup   int `json:"ruler_max_rules_per_rule_group"`
+	RulerMaxRuleGroupsPerTenant int `json:"ruler_max_rule_groups_per_tenant"`
+}
+
+// UserLimitsHandler handles user limits.
+func UserLimitsHandler(defaultLimits Limits, tenantLimits TenantLimits) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+
+		userID, err := tenant.TenantID(ctx)
```

Comment:
We don't need `ctx` elsewhere:

```suggestion
		userID, err := tenant.TenantID(r.Context())
```